Project/Assets/_Scripts/_new_Scripts/Weapons/RaycastGunViewModel.cs

```diff
@@ -2,123 +2,141 @@
+            model.BulletsInMagazine--;
 
-            StartCoroutine(ShotEffect());
+            for (int i = 0; i < model.BulletCount; ++i)
+            {
+                model.PenetrationDamage = model.Damage;
+                model.PenetrationLeft = model.BulletPenetration;
 
-            Vector3 direction = ProjectileSpawn.right;
-            direction.y += Random.Range(-model.BulletSpread.Value, model.BulletSpread.Value); // Creating bullet spread
+                StartCoroutine(ShotEffect());
 
-            Vector3 rayOrigin = ProjectileSpawn.position;
-            activeRay.SetPosition(0, rayOrigin);
+                Vector3 direction = ProjectileSpawn.right;
+                direction.y += Random.Range(-model.BulletSpread, model.BulletSpread); // Creating bullet spread
 
-            // Save every object on the ray's path and iterate over them to check how many were penetrated and damaged
-            RaycastHit[] hit;
-            hit = Physics.RaycastAll(rayOrigin, direction, model.RaycastLength.Value);
-            if (hit.Length > 0)
-            {
-                ShootOneRay(hit);
-            }
-            else
-            {
-                activeRay.SetPosition(1, rayOrigin + (direction * model.RaycastLength.Value));
+                Vector3 rayOrigin = ProjectileSpawn.position;
+                activeRay.SetPosition(0, rayOrigin);
+
+                // Save every object on the ray's path and iterate over them to check how many were penetrated and damaged
+                RaycastHit[] hit;
+                hit = Physics.RaycastAll(rayOrigin, direction, model.RaycastLength);
+                if (hit.Length > 0)
+                {
+                    ShootOneRay(hit);
+                }
+                else
+                {
+                    activeRay.SetPosition(1, rayOrigin + (direction * model.RaycastLength));
+                }
             }
         }
-    }
 
-    /// <summary>
-    /// Logic behind drawing ray, penetrating and dealing damage.
-    /// For example, if BulletCount = 5, then this function gets called 5 times
-    /// </summary>
-    private void ShootOneRay(RaycastHit[] hit)
-    {
-        for (int j = 0; j < hit.Length; ++j)
+        /// <summary>
+        /// Logic behind drawing ray, penetrating and dealing damage.
+        /// For example, if BulletCount = 5, then this function gets called 5 times
+        /// </summary>
+        private void ShootOneRay(RaycastHit[] hit)
         {
-            activeRay.SetPosition(1, hit[j].point);
-            Entity entity = hit[j].collider.GetComponent<Entity>();
-
-            if (entity != null)
+            for (int j = 0; j < hit.Length; ++j)
             {
-                entity.ReceiveDamage(model.PenetrationDamage.Value);
-                model.PenetrationDamage.Value -= entity.PenetrationDamageReduction;
+                activeRay.SetPosition(1, hit[j].point);
+                Entity entity = hit[j].collider.GetComponent<Entity>();
```

Comment:
Tego tu nie powinno być, powinniśmy mieć komponent IDamagable i tu go pobierać a następnie nadawać obrażenia.
Tak jak to jest w klasie CharacterViewModel i tam jest metoda damage. 

```csharp
public interface IDamagable 
{
        //true jeśli udało się obiekt zniszczyć
        bool Damage(float ammount);
}
```

Reply:
To poprawię dopiero w kolejnej wersji, trzeba obgadać jak Entity najpierw w ogóle zaimplementować, bo nie tylko postacie powinny go mieć i w ogóle jest klasą tymczasową